src/components/Form.js

```diff
@@ -0,0 +1,214 @@
+
+        /** Controls the loading state of the form */
+        isSubmitting: PropTypes.bool,
+
+        /** Server side error message */
+        serverErrorMessage: PropTypes.string,
+    }),
+
+    // eslint-disable-next-line react/forbid-prop-types
+    draftValues: PropTypes.object,
+
+    ...withLocalizePropTypes,
+};
+
+const defaultProps = {
+    formState: {
+        isSubmitting: false,
+        serverErrorMessage: '',
+    },
+    draftValues: {},
+};
+
+class Form extends React.Component {
+    constructor(props) {
+        super(props);
+
+        this.state = {
+            errors: {},
+        };
+
+        this.inputRefs = {};
+        this.touchedInputs = {};
+
+        this.getValues = this.getValues.bind(this);
+        this.setTouchedInput = this.setTouchedInput.bind(this);
+        this.validate = this.validate.bind(this);
+        this.submit = this.submit.bind(this);
+    }
+
+    /**
+     * @param {String} inputID - The inputID of the input being touched
+     */
+    setTouchedInput(inputID) {
+        this.touchedInputs[inputID] = true;
+    }
+
+    /**
+     * @returns {Object} - An object containing the values for each inputID
+     */
+    getValues() {
+        const values = {};
+        _.each(this.inputRefs, (inputRef, inputID) => {
+            values[inputID] = inputRef.value;
+        });
+        return values;
+    }
+
+    submit() {
+        // Return early if the form is already submitting to avoid duplicate submission
+        if (this.props.formState.isSubmitting) {
+            return;
+        }
+
+        // Touches all form inputs so we can validate the entire form
+        _.each(this.inputRefs, (inputRef, inputID) => (
+            this.touchedInputs[inputID] = true
+        ));
+
+        // Validate form and return early if any errors are found
+        const values = this.getValues();
+        if (!_.isEmpty(this.validate(values))) {
+            return;
+        }
+
+        // Set loading state and call submit handler
+        FormActions.setIsSubmitting(this.props.formID, true);
+        this.props.onSubmit(values);
+    }
+
+    /**
+     * @param {Object} values - An object containing the value of each inputID
+     * @returns {Object} - An object containing the errors for each inputID
+     */
+    validate(values) {
+        FormActions.setServerErrorMessage(this.props.formID, '');
+        const validationErrors = this.props.validate(values);
+        const errors = _.pick(validationErrors, (inputValue, inputID) => (
+            Boolean(this.touchedInputs[inputID])
+        ));
+        this.setState({errors});
+        return errors;
+    }
+
+    render() {
+        const childrenWrapperWithProps = children => (
+            // eslint-disable-next-line rulesdir/prefer-underscore-method
+            React.Children.map(children, (child) => {
+                // Just render the child if it is not a valid React element, e.g. text within a <Text> component
```

Comment:
Are there any other examples where we'd want to render things that aren't valid react elements?

Should we do this?

```js
const isValidReactElement = React.isValidElement(child);

if (!isValidReactElement) {
    return _.isString(child) && !_.isEmpty(child) ? child : null;
}
```

Reply:
Potentially functions returning a React Component and not element? But I don't recall seeing examples of this in the code. I'm not sure we would gain anything by strictly enforcing non-empty strings here. I think it's fine to keep it as is and just return/render anything that `Form` doesn't care about.